# PRCP-1027 – Skin Disorder Prediction Using Machine Learning

## 1. Problem Statement

Skin diseases are common healthcare problems that require early and accurate diagnosis. The differential diagnosis of erythemato‑squamous diseases is challenging because many diseases share similar clinical and histopathological features. Manual diagnosis often requires expert dermatologists and biopsy procedures, which are time‑consuming and subjective.

### Objectives of the Project

Task 1: Prepare a complete data analysis report on the given dermatology dataset.

Task 2: Build a machine learning predictive model to classify various skin diseases.

Task 3: Provide suggestions to doctors to identify skin diseases at the earliest stage.

### Domain: Healthcare

## 2. Dataset Description

The dermatology dataset contains patient clinical and histopathological data collected for diagnosing skin disorders.

Total Attributes: 34

Input Features: 33

Target Variable: class

Number of Classes: 6 skin diseases

Missing Values: Present in the Age attribute

### Disease Classes

Psoriasis

Seborrheic Dermatitis

Lichen Planus

Pityriasis Rosea

Chronic Dermatitis

Pityriasis Rubra Pilaris

## 3. Attribute Information

### Clinical Attributes 

Erythema, scaling, definite borders, itching, koebner phenomenon, polygonal papules, follicular papules, oral mucosal involvement, knee and elbow involvement, scalp involvement, family history (0 or 1).

### Histopathological Attributes

Melanin incontinence, eosinophils, PNL infiltrate, fibrosis, exocytosis, acanthosis, hyperkeratosis, parakeratosis, clubbing of rete ridges, elongation of rete ridges, thinning of suprapapillary epidermis, spongiform pustule, munro microabscess, focal hypergranulosis, disappearance of granular layer, basal layer damage, spongiosis, saw‑tooth appearance, follicular horn plug, perifollicular parakeratosis, inflammatory infiltrate, band‑like infiltrate.


### Age 

Represents patient age (linear value).

## 4. Step‑by‑Step Methodology / Algorithm

### Step 1: Import Required Libraries

NumPy, Pandas, Matplotlib, Seaborn, and Scikit‑learn libraries were used for data analysis, visualization, model building, and evaluation.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier


import warnings
warnings.filterwarnings('ignore')


### Step 2: Load the Dataset

The dataset was loaded into a Pandas DataFrame for further processing.

In [2]:
df = pd.read_csv("dataset_35_dermatology (1).csv")
df.head()


,erythema,scaling,definite_borders,itching,koebner_phenomenon,polygonal_papules,follicular_papules,oral_mucosal_involvement,knee_and_elbow_involvement,scalp_involvement,...,disappearance_of_the_granular_layer,vacuolisation_and_damage_of_basal_layer,spongiosis,saw-tooth_appearance_of_retes,follicular_horn_plug,perifollicular_parakeratosis,inflammatory_monoluclear_inflitrate,band-like_infiltrate,Age,class
0,2,2,0,3,0,0,0,0,1,0,...,0,0,3,0,0,0,1,0,55,2
1,3,3,3,2,1,0,0,0,1,1,...,0,0,0,0,0,0,1,0,8,1
2,2,1,2,3,1,3,0,3,0,0,...,0,2,3,2,0,0,2,3,26,3
3,2,2,2,0,0,0,0,0,3,2,...,3,0,0,0,0,0,3,0,40,1
4,2,3,2,2,2,2,0,2,0,0,...,2,3,2,3,0,0,2,3,45,3


### Step 3: Exploratory Data Analysis (EDA)

Checked dataset shape and structure

Analyzed data types and summary statistics

Identified missing values

Studied class distribution

In [3]:
df.info()
df.describe()
df.isnull().sum()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 366 entries, 0 to 365
Data columns (total 35 columns):
 #   Column                                    Non-Null Count  Dtype 
---  ------                                    --------------  ----- 
 0   erythema                                  366 non-null    int64 
 1   scaling                                   366 non-null    int64 
 2   definite_borders                          366 non-null    int64 
 3   itching                                   366 non-null    int64 
 4   koebner_phenomenon                        366 non-null    int64 
 5   polygonal_papules                         366 non-null    int64 
 6   follicular_papules                        366 non-null    int64 
 7   oral_mucosal_involvement                  366 non-null    int64 
 8   knee_and_elbow_involvement                366 non-null    int64 
 9   scalp_involvement                         366 non-null    int64 
 10  family_history                            366 non-

erythema                                    0
scaling                                     0
definite_borders                            0
itching                                     0
koebner_phenomenon                          0
polygonal_papules                           0
follicular_papules                          0
oral_mucosal_involvement                    0
knee_and_elbow_involvement                  0
scalp_involvement                           0
family_history                              0
melanin_incontinence                        0
eosinophils_in_the_infiltrate               0
PNL_infiltrate                              0
fibrosis_of_the_papillary_dermis            0
exocytosis                                  0
acanthosis                                  0
hyperkeratosis                              0
parakeratosis                               0
clubbing_of_the_rete_ridges                 0
elongation_of_the_rete_ridges               0
thinning_of_the_suprapapillary_epi

### Step 4: Data Preprocessing

Missing values in the Age column were replaced using the median value

Data was cleaned and prepared for modeling

In [4]:
(df == '?').sum()


erythema                                    0
scaling                                     0
definite_borders                            0
itching                                     0
koebner_phenomenon                          0
polygonal_papules                           0
follicular_papules                          0
oral_mucosal_involvement                    0
knee_and_elbow_involvement                  0
scalp_involvement                           0
family_history                              0
melanin_incontinence                        0
eosinophils_in_the_infiltrate               0
PNL_infiltrate                              0
fibrosis_of_the_papillary_dermis            0
exocytosis                                  0
acanthosis                                  0
hyperkeratosis                              0
parakeratosis                               0
clubbing_of_the_rete_ridges                 0
elongation_of_the_rete_ridges               0
thinning_of_the_suprapapillary_epi

In [5]:
df.replace('?', np.nan, inplace=True)

In [6]:
df['Age'] = pd.to_numeric(df['Age'])

In [7]:
df['Age'].fillna(df['Age'].median(), inplace=True)


### Step 5: Feature and Target Separation

Features (X): All columns except class

Target (y): class

In [8]:
X = df.drop('class', axis=1)
y = df['class']


### Step 6: Train‑Test Split

The dataset was divided into training (80%) and testing (20%) sets to evaluate model performance.

In [9]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)


### Step 7: Feature Scaling

StandardScaler was applied to normalize the feature values to improve model performance.

In [10]:
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)


### Step 8: Model Training

The following machine learning models were trained:

Logistic Regression

Decision Tree Classifier

Random Forest Classifier

Support Vector Machine (SVM)

K‑Nearest Neighbors (KNN)

In [11]:
lr = LogisticRegression(max_iter=1000)
lr.fit(X_train, y_train)
lr_pred = lr.predict(X_test)
print("Logistic Accuracy:", accuracy_score(y_test, lr_pred))


Logistic Accuracy: 0.9864864864864865


In [12]:
dt = DecisionTreeClassifier()
dt.fit(X_train, y_train)
dt_pred = dt.predict(X_test)
print("Decision Tree Accuracy:", accuracy_score(y_test, dt_pred))


Decision Tree Accuracy: 0.972972972972973


In [13]:
rf = RandomForestClassifier(n_estimators=100)
rf.fit(X_train, y_train)
rf_pred = rf.predict(X_test)

print("Random Forest Accuracy:", accuracy_score(y_test, rf_pred))
print(confusion_matrix(y_test, rf_pred))
print(classification_report(y_test, rf_pred))


Random Forest Accuracy: 0.9864864864864865
[[31  0  0  0  0  0]
 [ 0  9  0  0  0  0]
 [ 0  0 13  0  0  0]
 [ 0  1  0  7  0  0]
 [ 0  0  0  0 10  0]
 [ 0  0  0  0  0  3]]
              precision    recall  f1-score   support

           1       1.00      1.00      1.00        31
           2       0.90      1.00      0.95         9
           3       1.00      1.00      1.00        13
           4       1.00      0.88      0.93         8
           5       1.00      1.00      1.00        10
           6       1.00      1.00      1.00         3

    accuracy                           0.99        74
   macro avg       0.98      0.98      0.98        74
weighted avg       0.99      0.99      0.99        74



In [14]:
svm = SVC()
svm.fit(X_train, y_train)
svm_pred = svm.predict(X_test)
print("SVM Accuracy:", accuracy_score(y_test, svm_pred))


SVM Accuracy: 0.9864864864864865


In [15]:
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train, y_train)
knn_pred = knn.predict(X_test)
print("KNN Accuracy:", accuracy_score(y_test, knn_pred))


KNN Accuracy: 0.9864864864864865


### Step 9: Model Evaluation

Models were evaluated using:

Accuracy Score

Confusion Matrix

Classification Report

In [17]:
models = {
    "Logistic Regression": lr_pred,
    "Decision Tree": dt_pred,
    "Random Forest": rf_pred,
    "SVM": svm_pred,
    "KNN": knn_pred
}

In [21]:
final_results = []

for name, pred in models.items():
    report = classification_report(y_test, pred, output_dict=True)
    
    final_results.append({
        "Model": name,
        "Accuracy": accuracy_score(y_test, pred),
        "Precision (avg)": report["weighted avg"]["precision"],
        "Recall (avg)": report["weighted avg"]["recall"],
        "F1-score (avg)": report["weighted avg"]["f1-score"]
    })

final_df = pd.DataFrame(final_results)
final_df.sort_values(by="Accuracy", ascending=False)


,Model,Accuracy,Precision (avg),Recall (avg),F1-score (avg)
0,Logistic Regression,0.986486,0.987838,0.986486,0.986392
2,Random Forest,0.986486,0.987838,0.986486,0.986392
3,SVM,0.986486,0.987838,0.986486,0.986392
4,KNN,0.986486,0.987838,0.986486,0.986731
1,Decision Tree,0.972973,0.974897,0.972973,0.972852


### 5. Model Comparison – Accuracy Table

The performance of different machine learning models was evaluated using Accuracy, Precision, Recall, and F1-score to identify the most effective classifier for skin disorder prediction.

### Observation

Logistic Regression, Random Forest, SVM, and KNN all achieved the same high accuracy of 98.64% with similar precision, recall, and F1-scores. Since the performance is already very high and consistent, hyperparameter tuning is not required.

### 6. Confusion Matrix Analysis

The confusion matrix analysis indicates that most skin disease classes are correctly classified, with very few misclassifications.

High true positive rate across all classes

Minimal false positive and false negative predictions

Demonstrates strong generalization capability

These results confirm that the Logistic Regression model is reliable and effective for multi-class skin disease prediction.

### 7. Results and Discussion

The Logistic Regression model demonstrated strong overall performance and achieved the highest accuracy among the tested models. Its ability to handle overlapping class boundaries contributed to better classification results. Feature scaling significantly improved the performance of SVM and KNN models. Overall, Logistic Regression proved to be a robust and efficient model for this prediction task.

### 8. Suggestions to Doctors (Task 3)

Use AI-based prediction systems as a decision-support tool, not a replacement for clinical expertise

Identify skin disease patterns at early stages using patient symptoms and clinical features

Reduce dependency on invasive procedures such as biopsy during initial screening

Improve diagnostic speed and accuracy, leading to better patient care

### 9. Challenges Faced and Solutions
***Challenge 1: Missing Values***

Issue: Missing values in the Age column.
Solution: Median imputation was applied to handle missing values while minimizing the impact of outliers.

***Challenge 2: Similar Disease Features***

Issue: High similarity between different skin disease classes affected classification accuracy.
Solution: Logistic Regression was used due to its strong performance in handling overlapping class boundaries.

***Challenge 3: Feature Scaling***

Issue: Features were on different scales, which impacted model performance.
Solution: Feature scaling was performed using StandardScaler to normalize the data.

***Challenge 4: Model Selection***

Issue: Difficulty in selecting the best-performing model.
Solution: Multiple models were trained and evaluated using accuracy and other performance metrics to identify the optimal model.

### 10. Conclusion

A machine learning–based skin disorder prediction system was successfully developed and evaluated. Among all the models tested, the Logistic Regression model achieved the highest accuracy and demonstrated reliable performance. This system can assist dermatologists in early diagnosis, reduce diagnostic errors, and support informed healthcare decision-making.

### 11. Tools and Technologies Used

Python

NumPy

Pandas

Matplotlib & Seaborn

Scikit-learn

Jupyter Notebook

# End of Project